code/game/objects/items/devices/paicard.dm

```diff
@@ -52,99 +52,118 @@
-		for(var/slaws in pai.laws.supplied)
-			dat += "Additional directives: <br>[slaws]<br>"
-		dat += "<a href='byond://?src=[REF(src)];setlaws=1'>Configure Directives</a><br>"
-		dat += "<br>"
-		dat += "<h3>Device Settings</h3><br>"
-		if(pai.radio)
-			dat += "<b>Radio Uplink</b><br>"
-			dat += "Transmit: <A href='byond://?src=[REF(src)];toggle_transmit=1'>\[[pai.can_transmit? "Disable" : "Enable"] Radio Transmission\]</a><br>"
-			dat += "Receive: <A href='byond://?src=[REF(src)];toggle_receive=1'>\[[pai.can_receive? "Disable" : "Enable"] Radio Reception\]</a><br>"
-		else
-			dat += "<b>Radio Uplink</b><br>"
-			dat += "<font color=red><i>Radio firmware not loaded. Please install a pAI personality to load firmware.</i></font><br>"
-		if(ishuman(user))
-			var/mob/living/carbon/human/H = user
-			if(H.real_name == pai.master || H.dna.unique_enzymes == pai.master_dna)
-				dat += "<A href='byond://?src=[REF(src)];toggle_holo=1'>\[[pai.canholo? "Disable" : "Enable"] holomatrix projectors\]</a><br>"
-		dat += "<A href='byond://?src=[REF(src)];fix_speech=1'>\[Reset speech synthesis module\]</a><br>"
-		dat += "<A href='byond://?src=[REF(src)];wipe=1'>\[Wipe current pAI personality\]</a><br>"
-	else
-		dat += "No personality installed.<br>"
-		dat += "Searching for a personality... Press view available personalities to notify potential candidates."
-		dat += "<A href='byond://?src=[REF(src)];request=1'>\[View available personalities\]</a><br>"
-	user << browse(dat, "window=paicard")
-	onclose(user, "paicard")
-	return
-
-/obj/item/paicard/Topic(href, href_list)
+	ui_interact(user)
 
-	if(!usr || usr.stat)
-		return
+/obj/item/paicard/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PaiCard")
+		ui.open()
 
-	if(href_list["request"])
-		SSpai.findPAI(src, usr)
+/obj/item/paicard/ui_state(mob/user)
+	return GLOB.inventory_state
 
-	if(pai)
-		if(!(loc == usr))
-			return
-		if(href_list["setdna"])
+/obj/item/paicard/ui_data(mob/user)
+	. = ..()
+	var/list/data = list()
+	data["candidates"] = list()
+	if(!pai)
+		data["candidates"] = pool_candidates()
+		data["pai"] = null
+		return data
+	data["pai"] = list()
+	data["pai"]["can_holo"] = pai.canholo
+	data["pai"]["dna"] = pai.master_dna
+	data["pai"]["emagged"] = pai.emagged
+	data["pai"]["laws"] = pai.laws.supplied
+	data["pai"]["master"] = pai.master
+	data["pai"]["name"] = pai.name
+	data["pai"]["transmit"] = pai.can_transmit
+	data["pai"]["receive"] = pai.can_receive
+	return data
+
+/obj/item/paicard/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return FALSE
+	if(!usr || usr.stat)
+		return FALSE
+	if(loc != usr)
+		return FALSE
+	switch(action)
+		if("download")
+			var/datum/pai_candidate/candidate
+			for(var/datum/pai_candidate/checked_candidate as anything in SSpai.candidates)
+				if(params["key"] == checked_candidate.key)
+					candidate = checked_candidate
+			var/obj/item/paicard/card = src
```

Comment:
need to be doubly careful with this, the previous code ```card = locate(ref) in pai_cards_list``` wasn't working if I switched it to ```locate(src) in pai_cards_list```. I'm not sure if this will cause unexpected behavior, but it's working

Reply:
Why have this at all instead of just using `src` directly?

Reply:
src seems the more logical conclusion but the original code was locating it in pai_card_list for some inexplicable reason 
I didn't want to think I might be overlooking the reason in just using src